Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,112 @@
+  for example users have reported that they are using GitOps to fully manage the life cycle of Silences Custom resources,
+  which in fact brings benefits to the whole team because having a history and revisions in Git as well as constancy to manage the silences in a multi-tenant environment
+
+Additionally, having a new component in the stack and keeping it maintained is not always ideal (said the folks at [Giant Swarm](https://giantswarm.io) the owners of [silence-operator](https://github.com/giantswarm/silence-operator)),
+because that would require us to fully manage this component as of any managed app (monitoring, logging, alerting etc..).
+Having the Silence CRD as part of prometheus-operator provides users with a better experience: there is no need to install another operator or build a system on top of it to provide that functionality.
+
+By Adding support for `Silence` CRD in the prometheus-operator, this will make users more flexible in terms of choosing the tool
+to deploy their silences and will free others from managing a standalone component within the stack.
+
+## Pitfalls of the current solution
+
+Using Alertmanager API Directly comes with drawbacks:
+
+* Teams have to build an automation to add silences in a centralized manner
+* There is no input validation, which can lead to an invalid silence configuration
+* Teams need to manage expiration on their own
+* If a silence was deleted by accident, it will be permanent.
+
+# Goals
+
+* Provide a way for users to manage Alertmanager silences with Custom Resources (CR).
+* Enable management of Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
+* Multi tenancy: Silences CR can be using kubernetes rbac.
+* Moving the Giant Swarm silence operator to prometheus operator would provide better community support
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.
+* Users who want to manage silences the same way as for services running within the Kubernetes cluster.
+* Users who want a supported Kubernetes way of silences outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal do not cover automatic cleanup of expired Silence CRs
+* Refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+Creating a new cluster-scope Silence CRD that will act as an interface by adding silences via the Alertmanager API.
+
+A typical `Silence` resource could look like the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: Silence
+metadata:
+  name: my-silence
+  labels:
+    test: value
+spec:
+  selector:
+    app: test-alertmanager
+  expiresAt: "2023-06-01"
+  matchers:
+    - name: namespace
+      value: test-ns
+      isRegex: false
+```
+
+The above resource will result in creating a silence in the Alertmanager with label selector,
+`app: test-alertmanager`.
```

Comment:
I don't think that it's the correct model. This approach would allow any user to create silences on any Alertmanager in the cluster.
IMHO we need `.spec.silenceSelector` and `.spec.silenceNamespaceSelector` in the Alertmanager CRD.

Reply:
From my point of view, I dont think this is a good idea  `.spec.silenceSelector` and `.spec.silenceNamespaceSelector` in the Alertmanager CRD.

As silences are An API resource and it's not a config like `AlertmanagerConfig`.

Having the selector in the Silence CR, make it easier to select which Alertmanager to Use. 

What do you think?

Reply:
From a security point of view, I guess we can limit the users by RBAC who can Create/Update Silences.

Reply:
If you emphasize using this approach, then we need to use both selectors in Alertmanager CRD.
and remove this selector.

Reply:
I also thinks this makes more sense because silence creators should not be aware of which alertmanager they want to target

Reply:
This is an interesting problem! I agree with @simonpasquier but I also see the other side where it would be great if the user did not have to care about which AM they are targeting.
For the sake of discussion, if we were to only have `selector`, would it be feasible for the operator to only accept Silences in the same place as the `PrometheusRule` or would this be too limiting?

Reply:
We will just go with selectors in the alertmanager for now: 
```yaml
[...]
spec:
  silenceSelector: ...
  silenceNamespaceSelector: ...
```